superset/security/manager.py

```diff
@@ -2063,6 +2063,16 @@ def can_access_based_on_dashboard(datasource: "BaseDatasource") -> bool:
         )
 
         exists = db.session.query(query.exists()).scalar()
+        
+        # check for datasets that are only used by filters
+        if not exists:
+            dashboards_json = db.session.query(Dashboard.json_metadata).filter(Dashboard.id.in_(dashboard_ids)).all()
+            for json_ in dashboards_json:
+                json_metadata = json.loads(json_.json_metadata)
```

Comment:
This looks great! One small caution.. we don't have a lot of checks about the shape of this metadata, so I'd suggest putting a few try/catches  or data validation in here, first on the json.loads and then also below where you're accessing specific keys in the json that may/may not exist.